quickwit/quickwit-control-plane/src/indexing_plan.rs

```diff
@@ -0,0 +1,657 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::cmp::Ordering;
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use quickwit_cluster::ClusterMember;
+use quickwit_config::{SourceConfig, CLI_INGEST_SOURCE_ID, INGEST_API_SOURCE_ID};
+use quickwit_proto::indexing_api::IndexingTask;
+use serde::Serialize;
+
+/// A [`PhysicalIndexingPlan`] defines the list of indexing tasks
+/// each indexer, identified by its node ID, should run.
+/// TODO(fmassot): a metastore version number will be attached to the plan
+/// to identify if the plan is up to date with the metastore.
+#[derive(Debug, PartialEq, Clone, Serialize)]
+pub struct PhysicalIndexingPlan {
+    indexing_tasks_per_node_id: HashMap<String, Vec<IndexingTask>>,
+}
+
+impl PhysicalIndexingPlan {
+    /// Creates a [`PhysicalIndexingPlan`] from a non empty list of node IDs.
+    pub fn new(node_ids: Vec<String>) -> Self {
+        assert!(!node_ids.is_empty());
+        let indexing_tasks_per_indexer = node_ids
+            .into_iter()
+            .map(|node_id| (node_id, Vec::new()))
+            .collect();
+        Self {
+            indexing_tasks_per_node_id: indexing_tasks_per_indexer,
+        }
+    }
+
+    /// Adds the indexing task to the list of tasks assigned to the given node ID.
+    pub fn assign_indexing_task(&mut self, node_id: &String, indexing_task: IndexingTask) {
+        let node_indexing_tasks = self
+            .indexing_tasks_per_node_id
+            .get_mut(node_id)
+            .expect("The node must always be present.");
+        node_indexing_tasks.push(indexing_task);
+    }
+
+    /// Returns the number of indexing tasks for the given node ID.
+    pub fn num_indexing_tasks_for_node(&self, node_id: &String) -> usize {
+        self.indexing_tasks_per_node_id
+            .get(node_id)
+            .map(|tasks| tasks.len())
+            .unwrap_or(0)
+    }
+
+    /// Returns the number of indexing tasks for the given node ID, index ID and source ID.
+    pub fn num_indexing_tasks_for(&self, node_id: &str, index_id: &str, source_id: &str) -> usize {
+        self.indexing_tasks_per_node_id
+            .get(node_id)
+            .map(|tasks| {
+                tasks
+                    .iter()
+                    .filter(|task| task.index_id == index_id && task.source_id == source_id)
+                    .count()
+            })
+            .unwrap_or(0)
+    }
+
+    /// Returns the hashmap of (node ID, indexing tasks).
+    pub fn indexing_tasks_per_node(&self) -> &HashMap<String, Vec<IndexingTask>> {
+        &self.indexing_tasks_per_node_id
+    }
+
+    /// Returns the mean number of indexing tasks per node.
+    pub fn num_indexing_tasks_mean_per_node(&self) -> f32 {
+        let num_nodes = self.indexing_tasks_per_node_id.len();
```

Comment:
If you ditch the assert, make sure the `indexing_tasks_per_node_id` is non empty.